### PR TITLE
[FEATURE] add Fluent Datasources to `CloudDataContext`

### PR DESCRIPTION
**Pull request by Kilo59**
### Changes proposed in this pull request:
- Allow for Fluent Datasource to be added to and loaded from a `CloudDataContext`
- [add `responses` as testing dependency for more easily mocking cloud API endpoints](https://github.com/getsentry/responses)
   - [x] add this to requirements

### Followups
1. Delete Fluent Datasources
2. Persist assets immediately on `datasources.add_<ASSET_TYPE>()` calls.
3. Update Fluent Datasources

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

### CODE REVIEW

File: great_expectations/data_context/store/datasource_store.py
Comment (Kilo59) [inline great_expectations/data_context/store/datasource_store.py]: Should probably do this in the `Serializer`

File: great_expectations/data_context/store/datasource_store.py
Comment (Kilo59) [inline great_expectations/data_context/store/datasource_store.py]: Deal with these typing issues before merge
`_build_key_from_config` could accept a thinner interface/protocol.

File: great_expectations/data_context/store/inline_store_backend.py
Comment (Kilo59) [inline great_expectations/data_context/store/inline_store_backend.py:106]: When we call `self._datasource_store.set(key=None, value=datasource)` this causes the `fluent` datasources to be added to the `datasources` config entry, which will then be validated using the old "Datasource" marshmallow schemas.

This temporary hack is just to demonstrate the problem.

File: great_expectations/data_context/data_context/abstract_data_context.py
Comment (Kilo59) [inline great_expectations/data_context/data_context/abstract_data_context.py]: We don't seem to actually call this anywhere in our code right now.
The `context.sources.add_or_update_<DATASOURCE_TYPE>()` method always calls `_add_fluent_datasource`.

File: great_expectations/data_context/store/datasource_store.py
Comment (Kilo59) [inline great_expectations/data_context/store/datasource_store.py]: It seems natural to move this logic to the `serializer.serialize()` method, but the problem is that there's not an obvious way to move the `deserialization` away from this class as well.

Open to suggestions here.
I'd like to avoid changing too much at once if possible, my goal is to get this working so as to unblock the rest of Cloud FluentDatasources work, and followup with a better abstraction if it presents itself.

File: great_expectations/data_context/store/inline_store_backend.py
Comment (Kilo59) [inline great_expectations/data_context/store/inline_store_backend.py:106]: When we call `self._datasource_store.set(key=None, value=datasource)` this causes the `fluent` datasources to be added to the `"datasources"` config entry, which will then be validated using the old "Datasource" marshmallow schemas when we call `context.variables.save_config()`

This temporary hack is just to demonstrate the problem. 

The cleanest solution might be to also `stash` these values or simply remove fluent datasource types from `datasources` during the `_fluent_objects_stash()` call.

https://github.com/great-expectations/great_expectations/blob/2367d978b85ebec76c7e9bd81b6e562b0dc7f9c3/great_expectations/data_context/data_context_variables.py#L364-L398

File: great_expectations/data_context/store/inline_store_backend.py
Comment (Kilo59) [inline great_expectations/data_context/store/inline_store_backend.py]: When we call `self._datasource_store.set(key=None, value=datasource)` this causes the `fluent` datasources to be added to the `"datasources"` config entry, which will then be validated using the old "Datasource" marshmallow schemas when we call `context.variables.save_config()`

This temporary hack is just to demonstrate the problem. 

The cleanest solution might be to also `stash` these values or simply remove fluent datasource types from `datasources` during the `_fluent_objects_stash()` call.

https://github.com/great-expectations/great_expectations/blob/2367d978b85ebec76c7e9bd81b6e562b0dc7f9c3/great_expectations/data_context/data_context_variables.py#L364-L398

File: reqs/requirements-dev-lite.txt
Comment (cdkini) [inline reqs/requirements-dev-lite.txt:18]: If we go this route, can we make some noise so the team knows to use this moving forward?

File: great_expectations/data_context/store/datasource_store.py
Comment (cdkini) [reply great_expectations/data_context/store/datasource_store.py]: I think this is fine for now - as long as the branching is `FluentDatasource` vs `Datasource`, it doesn't feel like the worst violation. We should maybe add a ticket or brainstorm how to clean it up after the fact though

File: great_expectations/data_context/store/datasource_store.py
Comment (cdkini) [inline great_expectations/data_context/store/datasource_store.py]: Same note - if we use `type`, can we be clear that its the differentiating factor between `FluentDatasource` and `Datasource`

File: great_expectations/data_context/store/inline_store_backend.py
Comment (cdkini) [reply great_expectations/data_context/store/inline_store_backend.py]: this whole `InlineStoreBackend` class is quite messy tbh - I'm not sure what the stash flow would be like but we can discuss synchronously

File: great_expectations/data_context/store/datasource_store.py
Comment (Kilo59) [inline great_expectations/data_context/store/datasource_store.py:94]: assert the type is valid

File: great_expectations/data_context/store/datasource_store.py
Comment (Kilo59) [reply great_expectations/data_context/store/datasource_store.py]: Yes.
Also added an assert here as discussed.

File: great_expectations/data_context/store/inline_store_backend.py
Comment (Kilo59) [reply great_expectations/data_context/store/inline_store_backend.py]: The stash strategy won't work without additional refactoring. May revisit it in a followup.

File: great_expectations/data_context/store/datasource_store.py
Comment (cdkini) [inline great_expectations/data_context/store/datasource_store.py]: Should the `id_` type declaration go above the conditional?

File: great_expectations/data_context/store/inline_store_backend.py
Comment (cdkini) [inline great_expectations/data_context/store/inline_store_backend.py:106]: Do we have plans of removing this logic? Forgot where we landed

File: tests/datasource/fluent/test_viral_snippets.py
Comment (cdkini) [inline tests/datasource/fluent/test_viral_snippets.py:237]: Do we want to assert against the filesystem as well? We know the cache is updated but we want to make sure the store works as intended

File: great_expectations/data_context/store/inline_store_backend.py
Comment (Kilo59) [reply great_expectations/data_context/store/inline_store_backend.py:106]: Yes my goal is to remove this logic before resolving the ticket.

File: tests/datasource/fluent/test_viral_snippets.py
Comment (Kilo59) [reply tests/datasource/fluent/test_viral_snippets.py:237]: Yes, that's a good idea. I want to break that out into a separate test suite though since these tests are meant to be high level.

File: great_expectations/data_context/store/datasource_store.py
Comment (Kilo59) [reply great_expectations/data_context/store/datasource_store.py]: Done

File: tests/datasource/fluent/test_cloud.py
Comment (Kilo59) [inline tests/datasource/fluent/test_cloud.py]: @cdkini is there a better way to check that items have been added to the stores?
`empty_contexts.stores` was just an empty dict but I know that items are being added.

Comment (cdkini) [reply tests/datasource/fluent/test_cloud.py]: we could use a spy or use `store.get`?

Comment (Kilo59) [reply tests/datasource/fluent/test_cloud.py]: Yeah, will do one or both of those. Thanks

File: great_expectations/data_context/data_context/abstract_data_context.py
Comment (Kilo59) [inline great_expectations/data_context/data_context/abstract_data_context.py:720]: Temporary workaround while we update `stores` to work better with Fluent Datasources.
Without this we end up with duplicate entries for fluent datasources in both `"fluent_datasources"` and `"datasources"` config/yaml entries.

Comment (cdkini) [reply great_expectations/data_context/data_context/abstract_data_context.py:720]: Could we add this context as a comment above the block?

File: tests/datasource/fluent/conftest.py
Comment (cdkini) [inline tests/datasource/fluent/conftest.py:322]: Can you check with @billdirks to make sure this doesn't overlap with his work? I know he was looking for a proper Cloud mock so maybe y'all can work together

File: tests/datasource/fluent/test_contexts.py
Comment (cdkini) [inline tests/datasource/fluent/test_contexts.py:103]: What's the purpose of this in a test file?

**Review by cdkini — APPROVED** [review]

File: great_expectations/data_context/data_context/abstract_data_context.py
Comment (Kilo59) [reply great_expectations/data_context/data_context/abstract_data_context.py:720]: Will do.

File: tests/datasource/fluent/test_contexts.py
Comment (Kilo59) [reply tests/datasource/fluent/test_contexts.py:103]: I like to do it for 2 reasons.

1. It lets me run pytest for this test file by just running it with `python`. Which makes running the test from an IDE with `run python file` shortcut possible.
2. This is a very minor side benefit but it anchors the bottom of the file which makes git diffs much more readable.

File: great_expectations/data_context/data_context/abstract_data_context.py
Comment (billdirks) [reply great_expectations/data_context/data_context/abstract_data_context.py:720]: This seems to be syncing `datasource` with the returned `set_datasource`. It just happens that the only change is the id. Maybe this sync/copy method should live on datasource? This is fine for now, especially if it's changing soon per your comment.

File: great_expectations/data_context/store/datasource_store.py
Comment (billdirks) [inline great_expectations/data_context/store/datasource_store.py]: I think raising may be preferable to an assert here.

File: great_expectations/data_context/store/datasource_store.py
Comment (billdirks) [inline great_expectations/data_context/store/datasource_store.py]: `id_: str | None  = str(datasource_config.id) if datasource_config.id else datasource_config.id` may be more succinct.

File: great_expectations/data_context/store/inline_store_backend.py
Comment (billdirks) [inline great_expectations/data_context/store/inline_store_backend.py:103]: How do fluent datasources get added to the `datasources` key? Could we make this never happen?

File: great_expectations/util.py
Comment (billdirks) [inline great_expectations/util.py]: What's this ignore for?

File: tests/datasource/fluent/conftest.py
Comment (billdirks) [reply tests/datasource/fluent/conftest.py:322]: Maybe I'll be able to use this. Currently, I'm mocking our cloud store calls (such as get) which wrap calls to cloud so I'm mocking above the response dependency.

File: tests/datasource/fluent/conftest.py
Comment (billdirks) [inline tests/datasource/fluent/conftest.py:60]: My preference would be to move gx cloud specific fixtures and helpers into their own file. This isn't blocking though.

**Review by billdirks — COMMENTED** [review]
Thanks for doing this. This looks good. I haven't looked at tests super closely. My only real outstanding question is how to fluent datasources get added to the `datasources` (nonfluent) key. The answer may be outside this PR but I'm curious.

File: great_expectations/data_context/store/datasource_store.py
Comment (Kilo59) [reply great_expectations/data_context/store/datasource_store.py]: I'll raise `LookupError` or a `TypeError` here

File: great_expectations/data_context/store/datasource_store.py
Comment (Kilo59) [reply great_expectations/data_context/store/datasource_store.py]: Will do

File: great_expectations/data_context/store/inline_store_backend.py
Comment (Kilo59) [reply great_expectations/data_context/store/inline_store_backend.py:103]: Because of the `self._datasource_store.set(key=None, value=datasource)` call that's needed to invoke the CloudAPI `DatasourceStore` webservice calls on a `CloudDataContext`

@cdkini and I are working on making this not a problem and then we can remove this but it may require some not insignificant changes to our `Stores`. But this unblocks the rest of the cloud work while we make those changes.

File: great_expectations/util.py
Comment (Kilo59) [reply great_expectations/util.py]: mypy thinks there are overlapping overloads for this method.
I'll fix it or remove it from this PR.

File: tests/datasource/fluent/conftest.py
Comment (Kilo59) [reply tests/datasource/fluent/conftest.py:60]: Yeah if it's okay I'll make these changes in followup.

File: great_expectations/data_context/store/inline_store_backend.py
Comment (billdirks) [reply great_expectations/data_context/store/inline_store_backend.py:103]: Sounds good, thanks for the explanation.

**Review by billdirks — APPROVED** [review]
LGTM, thanks!

**Comment by Kilo59** [comment]
This will create a `great_expectations.yml` testing artifact if this bugfix is not merged first.
- https://github.com/great-expectations/great_expectations/pull/7657

**Pull request merged**
